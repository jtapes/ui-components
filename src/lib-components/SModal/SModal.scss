@import '../../theme/styles/variables';
@import '../../theme/styles/mixins';
@import '../../theme/styles/reset';
@import '../../theme/styles/fonts';
@import '../../theme/styles/global';

.s-modal-enter {
  .s-modal {
    &__inner {
      opacity: 0;
      transform: translateY(-100%);
    }

    &__close-icon {
      opacity: 0;
      transform: rotate(90deg);
    }

    &__backdrop {
      opacity: 0;
    }
  }
}

.s-modal-enter-to {
  transition: all 0.6s ease;

  .s-modal {
    &__inner {
      opacity: 1;
      transition: all ease 0.4s;
    }

    &__close-icon {
      opacity: 1;
      transform: rotate(0deg);
      transition: all ease 0.6s;
    }

    &__backdrop {
      opacity: 1;
      transition: all ease 0.4s;
    }
  }
}

.s-modal-leave-to {
  transition: all 0.4s ease;

  .s-modal {
    &__inner {
      opacity: 0;
      transform: translateY(-100%);
      transition: all ease 0.3s;
    }

    &__backdrop {
      opacity: 0;
      transition: all ease 0.4s;
    }
  }
}

.s-modal {
  @include z-index(modal);

  position: fixed;
  left: 0;
  top: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
  height: 100vh;

  &__inner {
    @include z-index(default);

    position: relative;
    display: flex;
    flex-direction: column;
    overflow: hidden;
    margin: $size--24;
    border-radius: $size--4;
    max-height: calc(100% - #{$size--24});
    background: $color--white;

    @include responsive(sm) {
      margin: 0;
      border-radius: 0;
      padding: 0;
      width: 100%;
      height: 100%;
      max-height: none;
    }
  }

  &__header {
    display: flex;
    align-items: center;
    margin-bottom: $size--32;
    padding: $size--24 $size--24 0 $size--24;
  }

  &__body {
    position: relative;
    overflow-y: scroll;
    padding: 0 $size--40 $size--32;
    width: 100%;
    max-height: 100vh;

    @include responsive(sm) {
      padding: 0 $size--24 $size--16;
    }
  }

  &__footer {
    padding: 0 $size--40 $size--32;
  }

  &__close {
    display: flex;
    flex-grow: 1;
    align-self: self-start;
    justify-content: flex-end;
    color: $color--font-secondary;
    transform-origin: center;
    transition: $animate--time;

    &-icon {
      cursor: pointer;
    }

    &:hover {
      color: $color--primary;
    }
  }

  &__backdrop {
    position: absolute;
    left: 0;
    top: 0;
    width: 100%;
    height: 100%;
    background: rgba(0, 0, 0, 0.5);
  }

  @include responsive(sm) {
    height: calc(100vh + #{$size--20});

    &__body {
      top: -$size--20;
    }
  }
}
