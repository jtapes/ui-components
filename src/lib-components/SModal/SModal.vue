<template>
  <transition name="s-modal">
    <div v-if="value" class="s-modal">
      <div class="s-modal__inner" :style="modalInnerStyle">
        <div class="s-modal__header">
          <slot name="header" />
          <div v-if="closable" class="s-modal__close">
            <svg-icon
              class="s-modal__close-icon icon-24"
              name="close-24-32"
              @click.native="close"
            />
          </div>
        </div>
        <div class="s-modal__body">
          <slot />
        </div>
        <div v-if="$slots.footer" class="s-modal__footer">
          <slot name="footer" />
        </div>
      </div>
      <div class="s-modal__backdrop" @click="close" />
    </div>
  </transition>
</template>

<script lang="ts">
import { Component, Model, Prop, Vue, Watch } from 'vue-property-decorator'
import SvgIcon from '@/lib-components/SvgIcon/SvgIcon.vue'

type ModalInnerStyle = {
  width?: string | number
  maxWidth?: string | number
}

@Component({
  components: {
    SvgIcon,
  }
})
export default class SModal extends Vue {
  /** Модель, отвечающая за видимость окна **/
  @Model('change')
  value!: boolean

  /** Ширина модального окна **/
  @Prop([String, Number])
  width?: string | number

  /** Максимальная ширина модального окна **/
  @Prop([String, Number])
  maxWidth?: string | number

  /** Максимальная ширина модального окна **/
  @Prop({ type: Boolean, default: true })
  closable!: boolean

  get modalInnerStyle(): ModalInnerStyle {
    const innerStyle: ModalInnerStyle = {}
    if (this.width) {
      if (Number.isInteger(this.width)) {
        innerStyle.width = `${this.width}px`
      } else {
        innerStyle.width = this.width
      }
    }

    if (this.maxWidth) {
      if (Number.isInteger(this.maxWidth)) {
        innerStyle.maxWidth = `${this.maxWidth}px`
      } else {
        innerStyle.maxWidth = this.maxWidth
      }
    }

    return innerStyle
  }

  close(): void {
    this.$emit('change', false)
  }

  @Watch('value', { immediate: true })
  handleValueChange(value: boolean): void {
    if (!this.$isServer) {
      document.body.style.overflow = value ? 'hidden' : 'auto'
    }
  }
}
</script>
