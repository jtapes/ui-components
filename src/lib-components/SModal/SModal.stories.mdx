import { action } from '@storybook/addon-actions'
import { config, withDesign } from 'storybook-addon-designs'
import { withTests } from '@storybook/addon-jest';
import {
  Meta,
  Story,
  ArgsTable,
  Description,
  Source,
  Canvas,
} from '@storybook/addon-docs'
import results from '/.jest-test-results.json';
import SModal from '@/lib-components/SModal/SModal.vue'

<Meta
  title="DESIGN SYSTEM/Atom/Modal"
  component={SModal}
  decorators={[withDesign, withTests({results})]}
  argTypes={{
  }}
  parameters={{
    design: config({
      type: 'figma',
      url: 'https://www.figma.com/file/IIeFkxNoHGcaqNHlVeNrhW/MVP-3-Market-UI-Kit?node-id=3019%3A1506',
    }),
    docs: {
      source: {
        type: 'code',
      },
    },
  }}
/>



export const Template = (args, {argTypes}) => ({
  props: Object.keys(argTypes),
  components: {SModal},
  template: `<div><button @click="openModal">Модальное окно</button><s-modal v-model="open" @close="closeModal">Контент</s-modal></div>`,
  data() {
    return {
      open: true
    }
  },
  methods: {
    closeModal() {
      action('close')()
      this.open = false
    },
    openModal() {
      this.open = true
    }
  },
})

<Canvas withToolbar withSource="none">
  <Story name="SModal">
    {Template.bind({})}
  </Story>
</Canvas>

<Description of={SModal} />
<ArgsTable of={SModal} />
<Source code={`<s-modal v-if="open" />`} />
