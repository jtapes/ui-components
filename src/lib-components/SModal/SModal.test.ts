import { mount } from '@vue/test-utils'
import SModal from '@/lib-components/SModal/SModal.vue'

describe('SModal', () => {
  const wrapper: any = mount(SModal, {
    slots: {
      default: 'Content',
    },
  })
  test('init', () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})
