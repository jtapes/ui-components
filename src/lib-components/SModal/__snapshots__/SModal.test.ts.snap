// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SModal init 1`] = `
<transition-stub
  name="s-modal"
>
  <!---->
</transition-stub>
`;
